lib/skate/detours/missed_stops.ex

```diff
@@ -0,0 +1,128 @@
+  end
+
+  defp missed_segments({connection_start, connection_end}, segmented_shape) do
+    %{index: start_index} = get_index_by_min_dist(segmented_shape, connection_start)
+    %{index: end_index} = get_index_by_min_dist(segmented_shape, connection_end)
+
+    if start_index == end_index do
+      %Skate.Detours.ShapeSegment{
+        # As long as there's only one stop in the segment,
+        # we can assume that we won't miss any stops
+        # because we're rejoining the same segment
+        stops: [_]
+      } = Enum.at(segmented_shape, start_index)
+
+      []
+    else
+      # We don't want the stop associated with the segment that the detour rejoins.
+      # So subtract one from the end index to ignore that segment
+      Enum.slice(segmented_shape, start_index..(end_index - 1))
+    end
+  end
+
+  defp segment_shape_by_stops([] = _shape, [] = _stops),
+    do: []
+
+  defp segment_shape_by_stops(shape, [] = _stops),
+    # If there are no stops, return the shape
+    do: [%Skate.Detours.ShapeSegment{points: shape, stops: []}]
+
+  defp segment_shape_by_stops(shape, stops) do
+    # Find the stop closest to the shape
+    [%{index: anchor_stop_index, shape_dist: [%{index: shape_point_anchor_index} | _]} | _] =
+      sort_stops_by_dist_to_shape(shape, stops)
+
+    # Split the shape and stops at the anchor point
+    # Add one to the Shape Point Anchor Index so that the point ends up in `left_shape`
+    {left_shape, right_shape} = Enum.split(shape, shape_point_anchor_index + 1)
+
+    {left_stops, [anchor | right_stops]} = Enum.split(stops, anchor_stop_index)
+
+    # Process `left_shape` into segments
+    {left_segments, [anchor_segment]} =
+      Enum.split(segment_shape_by_stops(left_shape, left_stops), -1)
+
+    # Take last segment, which should not have a stop,
+    # and reconnect it with the anchor stop
+    anchor_segment = %Skate.Detours.ShapeSegment{
+      (%Skate.Detours.ShapeSegment{stops: []} = anchor_segment)
+      | stops: [anchor]
+    }
+
+    left_segments ++ [anchor_segment] ++ segment_shape_by_stops(right_shape, right_stops)
+  end
+
+  defp sort_stops_by_dist_to_shape(shape, stops) do
+    stops
+    |> Enum.with_index(fn stop, index ->
+      %{
+        stop: stop,
+        index: index,
+        shape_dist: sort_by_distance(shape, stop)
+      }
+    end)
+    |> Enum.sort_by(fn %{shape_dist: [%{dist: dist} | _]} -> dist end)
+  end
+
+  defp get_index_by_min_dist(shape_segments, reference) do
+    shape_segments
+    |> Enum.with_index(fn %Skate.Detours.ShapeSegment{points: points} = segment, idx ->
+      points = sort_by_distance(points, reference)
+      # Pick closest points within fuzzy_dist=(0)
+      # Discard adjacent points
+      # Pick highest rank or return :indeterminate
+      %{segment: segment, sorted_points: points, index: idx}
+    end)
+    |> Enum.min_by(fn %{sorted_points: [%{dist: dist} | _]} -> dist end)
+  end
+
+  defp sort_by_distance(coordinates, reference) do
```

Comment:
A bunch of these calls to `sort_by` (maybe all of them?) are only being used to then retrieve the first item - that is, the minimum, per some criterion. I think it would be simpler to understand and computationally faster to do `min_by` instead, and replace even more of these arrays with single items.

Reply:
I have an approach for doing this kind of refactor that I generally quite like, and I would be happy to pair on it.

Reply:
Something to keep in mind, I've been working on a few "bugs" in the algorithim, and have quickly needed to use the rest of the values in these arrays, so I think we're going to run right back into this if we switch to `min_by`

Reply:
See uhh, this mess 😅

```
      [closest | next] = points

      points =
        ([closest] ++
           Enum.filter(next, fn %{dist: dist_from_reference} ->
             dist_from_reference - closest.dist < min_diff
           end))
        |> Enum.sort_by(& &1.index, :asc)
        |> Enum.chunk_while(
          [],
          fn
            element, [] ->
              {:cont, [element]}

            %{index: e_idx} = element, [%{index: acc_idx} | _] = acc ->
              case e_idx - acc_idx do
                1 -> {:cont, [element | acc]}
                0 -> throw(:duplicate_index)
                _ -> {:cont, acc, [element]}
              end
          end,
          fn acc -> {:cont, acc, []} end
        )
        |> Enum.map(&Enum.min_by(&1, fn %{dist: dist} -> dist end))

```

Reply:
Oh jeez!

Yeah okay maybe let's revisit once the 🐛 🐞 are squared away!

Reply:
a78eec61114f040b1e27dc69522db7933c0dc219